[MXNET-664] Support integer type in ImageIter

vandanavk opened the pull request:
## Description ##
Support int32 type for labels in ImageIter. By default, labels are of type float and therefore lead to precision issues.

## Checklist ##
### Essentials ###
- [ ] The PR title starts with [MXNET-664], 
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- [ ] To the my best knowledge, examples are not affected by this change

### Changes ###
- Added a new parameter to ImageIter that specifies the dtype of the label. By default, it is set to float32, which was the case in the old code.
- Modified test_imageiter() unit test to test the new parameter with int32 and float32.

## Comments ##
- RecordIO and ImageDetIter have not been changed.
- The example (profiler_imageiter.py) is unaffected

haojin2 commented on tests/python/unittest/test_image.py:
```diff
@@ -132,26 +132,30 @@ def test_color_normalize(self):
-            path_root='')
-        for batch in test_iter:
-            pass
+        for dtype in ['int32', 'float32']:
```
Does this only support int32 & float32? If possible please also check float64 and int64.

vandanavk replied:
Sure, I can check float64 and int64

vandanavk commented:
@sandeep-krishnamurthy @apeforest @Roshrini

apeforest commented on python/mxnet/image/image.py:
```diff
@@ -1091,7 +1093,7 @@ def __init__(self, batch_size, data_shape, label_width=1,
                 for line in iter(fin.readline, ''):
                     line = line.strip().split('\t')
-                    label = nd.array([float(i) for i in line[1:-1]])
+                    label = nd.array([i for i in line[1:-1]], dtype=dtype)
```
[i for i in line[1:-1]] is just line[1:-1], right?

vandanavk replied:
Correct. made the change

apeforest commented on tests/python/unittest/test_image.py:
```diff
@@ -132,26 +132,30 @@ def test_color_normalize(self):
-        test_iter = mx.image.ImageIter(2, (3, 224, 224), label_width=1, imglist=im_list,
-            path_root='')
-        for _ in range(3):
+        def check_ImageIter(dtype='float32'):
```
nit: use lower case imageiter to be consistent with test_imageiter

vandanavk replied:
Done

sandeep-krishnamurthy commented on python/mxnet/image/image.py:
```diff
@@ -1091,7 +1093,7 @@ def __init__(self, batch_size, data_shape, label_width=1,
                 for line in iter(fin.readline, ''):
                     line = line.strip().split('\t')
-                    label = nd.array([float(i) for i in line[1:-1]])
+                    label = nd.array(line[1:-1], dtype=dtype)
```
What if I pass dtype='int8' or something that is not supported?

vandanavk commented:
@sandeep-krishnamurthy I've updated the commit - addressed your review comments. Please have a look at the updated changes and let me know your inputs.

Roshrini commented:
Thanks for working on this. LGTM.

vandanavk commented:
Thanks @haojin2 @apeforest @sandeep-krishnamurthy @Roshrini 

@szha Can this change be merged?
